[SYCL][ESIMD] Add test on simd broadcast constructor

vasilytric opened the pull request:
- In this PR was used utility-function for obtaining reference data. It lets avoid code copy-pasteing
- Was added function for obtaining reference data that depends on two data types

vasilytric committed:
[SYCL][ESIMD] Add test on simd broadcast constructor

vasilytric committed:
[SYCL][ESIMD] Add utility function for constructing ref data

It removes duplicated code

vasilytric committed:
[SYCL][ESIMD] Add function for generate converted reference data

This function provides reference data in the tests that uses a few data
types.

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_broadcast.hpp:
```diff
@@ -0,0 +1,196 @@
+
+namespace esimd = sycl::ext::intel::experimental::esimd;
+
+namespace esimd_test::api::functional {
```
```suggestion
namespace esimd_test::api::functional::ctors {
```
What do you think?

vasilytric replied:
done in this commit: b12bfe332f0da75f23cdd49b29d10318fd4961db

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_broadcast.hpp:
```diff
@@ -0,0 +1,196 @@
+  static std::string get_description() { return "initializer"; }
+
+  template <typename SrcT, typename DstT, int NumElems>
+  static void call_simd_ctor(DstT ref_data, SrcT *const out) {
```
```suggestion
  static void call_simd_ctor(DstT ref_value, SrcT *const out) {
```
It's more like a single value than a whole reference data generated. What do you think?

vasilytric replied:
`ref_value` mean that we have a single value, and in other tests we have `ref_data`

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_broadcast.hpp:
```diff
@@ -0,0 +1,196 @@
+  static std::string get_description() { return "variable declaration"; }
+
+  template <typename SrcT, typename DstT, int NumElems>
+  static void call_simd_ctor(DstT ref_data, SrcT *const out) {
+    const esimd::simd<SrcT, NumElems> simd_by_var_decl(ref_data);
```
```suggestion
  static void call_simd_ctor(SrcT ref_data, DstT *const out) {
    const esimd::simd<DstT, NumElems> simd_by_var_decl(ref_data);
```
The source and destination naming is a bit confusing over the test code. Is there anything I am missing here?

vasilytric replied:
done in this commit: b12bfe332f0da75f23cdd49b29d10318fd4961db

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_broadcast.hpp:
```diff
@@ -0,0 +1,196 @@
+                  const std::string &dst_data_type) {
+    bool passed = true;
+    const std::vector<SrcT> ref_data =
+        generate_converted_ref_data<SrcT, DstT, 1>();
```
```suggestion
        generate_ref_conv_data<SrcT, DstT, 1>();
```
or
```suggestion
        generate_ref_data_for_conversion<SrcT, DstT, 1>();
```

vasilytric replied:
done. `generate_ref_conv_data` the current name

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_broadcast.hpp:
```diff
@@ -0,0 +1,196 @@
+};
+
+template <typename DstT, int NumElems, typename ContextT>
+class ConvCtorTestDescription : public ITestDescription {
```
Do you intend to re-use it for converting constructor?

vasilytric replied:
I think no, so let's make decision when these tests will be merged

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_broadcast.hpp:
```diff
@@ -0,0 +1,196 @@
+  const size_t m_index;
+};
+
+template <typename, int, typename> class Kernel;
```
```suggestion
template <typename, int, typename...> class Kernel;
```
How do you think, would something like this work for every test in `functional/ctors`? So we might just update the `Kernel` in `common.hpp`.

vasilytric replied:
done in this commit: 1c2fb83daefec26bfb7049024ff22b780f794dc6

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_broadcast.hpp:
```diff
@@ -0,0 +1,196 @@
+
+    bool passed = true;
+    for (size_t i = 0; i < result.size(); ++i) {
+      if constexpr (std::is_same_v<SrcT, DstT>) {
+        if (!are_bitwise_equal(ref_val, result[i])) {
+          fail_test(i, result[i], ref_val, src_data_type, dst_data_type);
+        }
+      } else {
+        DstT expected = static_cast<DstT>(ref_val);
+        DstT retrieved = result[i];
```
```suggestion
      const DstT& expected = static_cast<DstT>(ref_val);
      const DstT& retrieved = result[i];
      if constexpr (std::is_same_v<SrcT, DstT>) {
        if (!are_bitwise_equal(expected, retrieved)) {
          passed = fail_test(i, retrieved, expected, src_data_type, dst_data_type);
        }
      } else  if constexpr (...) {
```

vasilytric replied:
done in this commit: b12bfe332f0da75f23cdd49b29d10318fd4961db

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_broadcast_core.cpp:
```diff
@@ -0,0 +1,50 @@
+  const auto types = get_tested_types<tested_types::fp>();
+  const auto single_dim = values_pack<8>();
+
+  // Run for specific combinations of types, vector length, base and step values
+  // and invocation contexts.
+  run_test<float, 8, double, initializer>{}(queue, "float", "double");
```
It's a bit unclear which one is the source and which one is destination type

vasilytric replied:
done [here](https://github.com/intel/llvm-test-suite/pull/690/commits/8b00738f82d597bacc82aa5c400cd503a2da80cc#diff-e3103512a9fb456b0e983db0746b4c64e4ae20d76e626d6ae2eb93293aa0056dR48-R50)

vasilytric committed:
[SYCL][ESIMD] Update broadcast ctor tests

- Remove using ctor namespace, use "ctor::" insted
- Fix logic with definin type in simd vector
- Change test result verification logic

vasilytric committed:
Merge branch 'intel' into simd_ctor_broadcast

vasilytric committed:
[SYCL][ESIMD] Add is_sycl_signed

Because std::is_signed can't handle sycl::half

vasilytric committed:
[SYCL][ESIMD] Update test with using for_all_combinations

vasilytric committed:
[SYCL][ESIMD] Update test logic and small changes

- Add comments

vasilytric committed:
[SYCL][ESIMD] Add a stub values to value_conv

vasilytric committed:
Merge branch 'intel' into simd_ctor_broadcast

vasilytric committed:
Merge branch 'intel' into simd_ctor_broadcast

vasilytric committed:
[SYCL][ESIMD] Apply clang-format

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_broadcast.hpp:
```diff
@@ -0,0 +1,210 @@
+
+    log_msg += m_src_data_type + ", " + std::to_string(NumElems) + ">";
+    log_msg += ", with context: " + ContextT::get_description();
+    log_msg += ", with source value type: " + m_src_data_type;
+    log_msg += ", with destination value type: " + m_dst_data_type;
+    log_msg += ", retrieved: " + std::to_string(m_retrieved_val);
+    log_msg += ", expected: " + std::to_string(m_expected_val);
+    log_msg += ", ref value: " + std::to_string(m_ref_val);
```
```suggestion
    log_msg += ", source type: " + m_src_data_type;
    log_msg += ", destination type: " + m_dst_data_type;
    log_msg += ", retrieved: " + std::to_string(m_retrieved_val);
    log_msg += ", expected: " + std::to_string(m_expected_val);
    log_msg += ", input: " + std::to_string(m_ref_val);
```

Also, as a suggestion.
So, currently we have the following message, for example:
`Failed for simd<float,8>, with context: variable declaration, with source value type: float, with destination value type: int, retrieved: 42, expected: 41, ref value: 41.5, at index: 0`

What if we have something like this instead:
`Failed for simd<float,8>, with context: variable declaration, destination type: int. Unexpected value at index 0. Input: 41.5, retrieved: 42, expected: 41`

yuriykoch replied:
Whoops, one more thing. I see the following code:
```
  template <typename SrcT, typename DstT, int NumElems>
  static void call_simd_ctor(SrcT ref_value, DstT *const out) {
    const auto simd_by_init = esimd::simd<DstT, NumElems>(ref_value);
    simd_by_init.copy_to(out);
  }
```
Is using `"Failed for simd<" +  m_src_data_type + ", " + std::to_string(NumElems) + ">"` valid here?

vasilytric replied:
> What if we have something like this instead:

No, I'd like to provide source and destination types

vasilytric replied:
this comment applied in this commit: da1b71a5bd6811a44a5ea48d03bb911518fe1655

yuriykoch replied:
Shouldn't we use `"Failed for simd<" + m_dst_data_type`?

vasilytric replied:
this comment was applied [here](https://github.com/intel/llvm-test-suite/pull/690/commits/be308ab37a2211afe7f091450cf5fb97c3e6346e#diff-056299db9fa515e78e72e7973d8a03b15b2a0ddb3557d03188e25a7b9b6edae1L97-R97)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_broadcast.hpp:
```diff
@@ -0,0 +1,210 @@
+
+// The main test routine.
+// Using functor class to be able to iterate over the pre-defined data types.
+template <typename UsePositiveValueOnly, typename SrcT, typename DimT,
```
```suggestion
template <typename UsePositiveValueOnly, typename SrcT, typename SizeT,
```

vasilytric replied:
this comment applied in this commit: https://github.com/intel/llvm-test-suite/pull/690/commits/d24d26838de20c84064ceadb145c5703b9c96ad4

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_broadcast_core.cpp:
```diff
@@ -0,0 +1,80 @@
+  // The source types is the first types, that provided to the
+  // "for_all_combinations" the destination types is the second types that
+  // provided to the "for_all_combinations".
+  passed &= for_all_combinations<ctors::run_test, use_ref_conv_values>(
+      int_type, two_dims, uint_types, all_contexts, queue);
+  passed &= for_all_combinations<ctors::run_test, use_positive_value_only>(
+      core_types, all_dims, core_types, all_contexts, queue);
```
Is there any possibility we can avoid using `use_ref_conv_values` explicitly every time?

vasilytric replied:
no in current implementation `use_ref_conv_values` is provides as the first argument to the `ctors::run_test`, but if  `use_ref_conv_values` provided to the last argument, then my answer will be changed to "yes" on your question

yuriykoch replied:
What about type aliases or tag dispatching?
```
namespace detail {
template <bool UsePositiveValueOnly, TemplateArgs>
struct run_test_impl {
  // Previously run_test
};
} // namespace detail

template <TemplateArgs>
using run_test = detail::run_test_impl<false, TemplateArgs>;

template <TemplateArgs>
using run_test_with_positive_values_only = detail::run_test_impl<true, TemplateArgs>;

....

passed &= for_all_combinations<ctors::run_test>(
      int_type, two_dims, uint_types, all_contexts, queue);
passed &= for_all_combinations<ctors::run_test_with_positive_values_only>(
      core_types, all_dims, core_types, all_contexts, queue);
```
My concern is that a corner case affects the verbosity of the main path, and the naming of the `use_ref_conv_values` depends on the name of `value_conv.hpp :: generate_ref_conv_data()`

vasilytric replied:
this comment applied in this commit: https://github.com/intel/llvm-test-suite/pull/690/commits/686c1e20d75c98859ac709d4769f35c95bcfb2de

yuriykoch replied:
How does using `ctors::run_test_with_ref_conv_values` solve concerns above?

vasilytric replied:
name was changed [here](https://github.com/intel/llvm-test-suite/pull/690/commits/be308ab37a2211afe7f091450cf5fb97c3e6346e#diff-056299db9fa515e78e72e7973d8a03b15b2a0ddb3557d03188e25a7b9b6edae1R211)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_broadcast.hpp:
```diff
@@ -0,0 +1,210 @@
+  call_simd_by_const_ref(const esimd::simd<DstT, NumElems> &simd_by_const_ref,
+                         DstT *const out) {
+    simd_by_const_ref.copy_to(out);
+  }
```
```suggestion
  template <typename DstT, int NumElems>
  static void
  call_simd_by_const_ref(const esimd::simd<DstT, NumElems> &simd_by_const_ref,
                         DstT *const out) {
    simd_by_const_ref.copy_to(out);
  }
```

vasilytric replied:
this comment applied [here](https://github.com/intel/llvm-test-suite/pull/690/commits/da1b71a5bd6811a44a5ea48d03bb911518fe1655#diff-056299db9fa515e78e72e7973d8a03b15b2a0ddb3557d03188e25a7b9b6edae1L68-R75)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_broadcast_fp_extra.cpp:
```diff
@@ -0,0 +1,60 @@
+  const auto fp_extra_types = get_tested_types<tested_types::fp_extra>();
+  using use_ref_conv_values = std::false_type;
+  const auto single_dim = get_dimensions<8>();
+  const auto vardecl_context = unnamed_type_pack<ctors::var_decl>::generate();
```
```suggestion
  const auto context = unnamed_type_pack<ctors::var_decl>::generate();
```
Nit: I would suggest to make names a bit more generic. Otherwise how many renames will you do if you select a different value to store?

vasilytric replied:
this comment applied [here](https://github.com/intel/llvm-test-suite/pull/690/commits/da1b71a5bd6811a44a5ea48d03bb911518fe1655#diff-e3103512a9fb456b0e983db0746b4c64e4ae20d76e626d6ae2eb93293aa0056dR44)

yuriykoch commented on SYCL/ESIMD/api/functional/type_traits.hpp:
```diff
@@ -31,7 +31,7 @@ inline constexpr bool is_sycl_floating_point_v{
 template <typename T>
 using is_sycl_signed =
-    std::bool_constant<std::is_signed_v<T> || is_sycl_floating_point_v<T>>;
+    std::bool_constant<std::is_signed_v<T> || std::is_same_v<T, sycl::half>>;
```
What if we would have some additional "sycl floating point" types?

vasilytric replied:
and if they will be signed types this statement should be extended

yuriykoch replied:
I see no possibility for floating point types to be unsigned, actually.
What's the reason behind the change anyway?

vasilytric replied:
ok, it will be reverted

vasilytric replied:
this will be reverted [here](https://github.com/intel/llvm-test-suite/pull/690/commits/be308ab37a2211afe7f091450cf5fb97c3e6346e#diff-7b267ff200ef7dbfe5e2b1fe912dcbc7a8622f4d14a995f2f683282cff5e1b49R34)

yuriykoch commented on SYCL/ESIMD/api/functional/value_conv.hpp:
```diff
@@ -51,15 +52,20 @@ std::vector<SrcT> generate_ref_conv_data() {
-  static const SrcT denorm = value_conv<SrcT, DstT>::denorm_min();
   static const SrcT positive = static_cast<SrcT>(126.75);
+  static const SrcT max = 10;
+  static const SrcT min = 2;
```
```suggestion
  static const SrcT min = std::min(-max, 0); // Use zero for unsigned types
```

vasilytric replied:
I suggest do not change these values because this is a temporary values that will be replaced with tested values with avoiding UB

vasilytric replied:
I've done it in this commit: 82a0ccf5343c935145775266bfea76c88c191fbd

vasilytric committed:
Merge branch 'intel' into simd_ctor_broadcast

vasilytric committed:
[SYCL][ESIMD] Structure changes

- Change error output
- Rename variables

vasilytric committed:
[SYCL][ESIMD] Replace "dim" with "size"

vasilytric committed:
[SYCL][ESIMD] Use an aliases instead of provide extra argument

vasilytric committed:
[SYCL][ESIMD] Update min value for converted reference data

vasilytric committed:
[SYCL][ESIMD] Apply clang-format

vasilytric committed:
[SYCL][ESIMD] Update converted reference data min value

vasilytric committed:
[SYCL][ESIMD] Update error message, rename variable

yuriykoch commented on SYCL/ESIMD/api/functional/value_conv.hpp:
```diff
@@ -70,7 +78,8 @@ std::vector<SrcT> generate_ref_conv_data() {
     ref_data = details::construct_ref_data<SrcT, NumElems>(
         {-0.0, max, max_half, min, min_half});
-  } else if constexpr (std::is_signed_v<SrcT> && std::is_signed_v<DstT>) {
+  } else if constexpr (std::is_signed_v<SrcT> &&
+                       type_traits::is_sycl_signed_v<DstT>) {
```
Is it intended? Please add comment if so
```suggestion
  } else if constexpr (type_traits::is_sycl_signed_v<SrcT> &&
                       type_traits::is_sycl_signed_v<DstT>) {
```

vasilytric replied:
I've applied this comment in this commit: 474bee149e79fe6831f55e81dae33379d914e58c

yuriykoch commented on SYCL/ESIMD/api/functional/value_conv.hpp:
```diff
@@ -70,7 +78,8 @@ std::vector<SrcT> generate_ref_conv_data() {
+                       type_traits::is_sycl_signed_v<DstT>) {
     ref_data = details::construct_ref_data<SrcT, NumElems>(
         {min, min_half, 0, max_half, max});
   } else if constexpr (std::is_signed_v<SrcT> && std::is_unsigned_v<DstT>) {
```
```suggestion
  } else if constexpr (type_traits::is_sycl_signed_v<SrcT> && std::is_unsigned_v<DstT>) {
```

vasilytric replied:
I've applied this comment in this commit: 474bee149e79fe6831f55e81dae33379d914e58c

yuriykoch commented on SYCL/ESIMD/api/functional/value_conv.hpp:
```diff
@@ -51,15 +52,22 @@ std::vector<SrcT> generate_ref_conv_data() {
-  static const SrcT inf = value<SrcT>::inf();
-  static const SrcT denorm = value_conv<SrcT, DstT>::denorm_min();
   static const SrcT positive = static_cast<SrcT>(126.75);
+  static const SrcT max = 10;
+  // Use zero for unsigned types
+  static const SrcT min =
+      std::min(static_cast<SrcT>(-max), static_cast<SrcT>(0));
```
```suggestion
  static const SrcT max = 10;
  // Use zero for unsigned types
  static const SrcT min =
      std::min(static_cast<SrcT>(-max), 0);
```

vasilytric replied:
when I try to apply this comment I got the following compilation error:
```
error: no matching function for call to 'min'
      std::min(static_cast<SrcT>(-max), 0);
note: candidate template ignored: deduced conflicting types for parameter '_Ty' ('float' vs. 'int')
_NODISCARD _Post_equal_to_(_Right < _Left ? _Right : _Left) constexpr const _Ty&(min)(
                                                                                 ^
```

yuriykoch replied:
```suggestion
  static const SrcT max = 10;
  // Use zero for unsigned types
  static const SrcT min = std::min<SrcT>(-max, 0);
```

vasilytric replied:
this comment was applied in this commit: 33d800706e8c6edfd581023acf7402823d73d8c6

vasilytric committed:
[SYCL][ESIMD] Replace std::is_signed_v with type_traits::is_sycl_signed_v

vasilytric committed:
Merge branch 'intel' into simd_ctor_broadcast

vasilytric committed:
[SYCL][ESIMD] update converted value min

vasilytric commented:
@sndmitriev Would you mind to review this PR?

vasilytric commented:
@kbobrovs Would you mind to merge this PR?
